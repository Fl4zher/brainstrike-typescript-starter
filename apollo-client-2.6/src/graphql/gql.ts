import gql from 'graphql-tag';

export const CARD_DATA = gql`
  fragment CardData on Card {
    __typename
    created
    updated
    id
    label
    number
  }
`;

export const GET_CARD_DATA = gql`
  query getCards(
    $first: Int
    $last: Int
    $after: String
    $before: String
    $orderByColumn: String
    $orderByDirection: DirectionEnum
    $categoryId: ID
  ) {
    cards(
      first: $first
      last: $last
      after: $after
      before: $before
      orderByColumn: $orderByColumn
      orderByDirection: $orderByDirection
      categoryId: $categoryId
    ) @connection(key: "Card_card") {
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
        totalCount
      }
      edges {
        cursor
        node {
          ...CardData
        }
      }
    }
  }
  ${CARD_DATA}
`;

export const GET_CARD_CATEGORIES = gql`
  query getCategories {
    categories {
      id
      name
      parentId
      created
      updated
    }
  }
`;
