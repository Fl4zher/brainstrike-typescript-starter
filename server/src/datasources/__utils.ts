export interface FindInterface<Entity> {
  after?: string;
  /**
   * Limit (paginated) - max number of entities should be taken.
   */
  first?: number;
  /**
   * Order, in which entities should be ordered.
   */
  sortOptions?: {
    [P in keyof Entity]?: "ASC" | "DESC" | 1 | -1;
  };
}
