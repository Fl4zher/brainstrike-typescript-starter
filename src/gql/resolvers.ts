import { merge } from "lodash";
import { GraphQLDate, GraphQLTime, GraphQLDateTime } from "graphql-iso-date";

import cardResolvers from "./card/resolvers";
import userResolvers from "./user/resolvers";

// Default Resolvers
const resolvers = {
  Date: GraphQLDate,
  Time: GraphQLTime,
  DateTime: GraphQLDateTime
};

export default merge(resolvers, cardResolvers, userResolvers);
