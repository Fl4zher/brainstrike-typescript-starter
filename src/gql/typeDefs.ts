import { importSchema } from "graphql-import";

export const typeDefs = importSchema(
  "src/gql/schema.graphql"
); /* Warning: Must be an absolute path */

export default typeDefs;
