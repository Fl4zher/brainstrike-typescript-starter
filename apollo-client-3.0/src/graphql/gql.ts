import gql from 'graphql-tag';

export const NODE_FRAGMENT = gql`
  fragment NodeFragment on Node {
    created
    updated
  }
`;

export const CARD_FRAGMENT = gql`
  fragment CardFragment on Card {
    id
    ...NodeFragment
    __typename
    label
    number
  }
  ${NODE_FRAGMENT}
`;

export const GET_CARD_DATA = gql`
  query getCards(
    $first: Int
    $last: Int
    $after: String
    $before: String
    $orderByColumn: String
    $orderByDirection: DirectionEnum
    $categoryId: ID
  ) {
    cards(
      first: $first
      last: $last
      after: $after
      before: $before
      orderByColumn: $orderByColumn
      orderByDirection: $orderByDirection
      categoryId: $categoryId
    ) @connection(key: "Cards") {
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
        totalCount
      }
      edges {
        cursor
        node {
          ...CardFragment
        }
      }
    }
  }
  ${CARD_FRAGMENT}
`;

export const CATEGORY_FRAGMENT = gql`
  fragment CategoryFragment on Category {
    __typename
    id
    created
    updated
    name
    parentId
  }
`;

export const GET_CARD_CATEGORIES = gql`
  query getCategories {
    categories {
      ...CategoryFragment
    }
  }
  ${CATEGORY_FRAGMENT}
`;

export const GET_CATEGORY = gql`
  query getCategory(
    $id: ID!
    $first: Int
    $last: Int
    $after: String
    $before: String
    $orderByColumn: String
    $orderByDirection: DirectionEnum
  ) {
    category: node(id: $id) {
      ... on Node {
        id
        created
        updated
      }
      ... on Category {
        __typename
        name
        parentId
        _cards(
          first: $first
          last: $last
          after: $after
          before: $before
          orderByColumn: $orderByColumn
          orderByDirection: $orderByDirection
        ) {
          pageInfo {
            hasNextPage
            hasPreviousPage
            startCursor
            endCursor
            totalCount
          }
          edges {
            cursor
            node {
              __typename
              id
              label
              number
              created
              updated
            }
          }
        }
      }
    }
  }
`;
