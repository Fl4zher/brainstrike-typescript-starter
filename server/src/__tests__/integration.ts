import { createTestClient } from "apollo-server-testing";
import gql from "graphql-tag";
import { constructTestServer, createDbConnection, Connection } from "./__utils";
import {
  mockFirstCardResponse,
  mockFirstCardResponseId,
  mockReturnCard,
  mockCardInput
} from "../datasources/__tests__/card";

const GET_CARDS = gql`
  query getCards {
    cards {
      id
      number
      label
      description
    }
  }
`;

const GET_CARD = gql`
  query card($id: ID!) {
    card(id: $id) {
      id
      number
      label
      description
    }
  }
`;

const ADD_CARD = gql`
  mutation addCard($input: CardInput!) {
    addCard(input: $input) {
      success
      message
      card {
        id
        number
        label
        description
      }
    }
  }
`;

const UPDATE_CARD = gql`
  mutation updateCard($id: ID!, $input: CardInput!) {
    updateCard(id: $id, input: $input) {
      success
      message
      card {
        id
        number
        label
        description
      }
    }
  }
`;

const REMOVE_CARD = gql`
  mutation removeCard($id: ID!) {
    removeCard(id: $id) {
      success
      message
    }
  }
`;

let connection: Connection;

beforeAll(async () => {
  console.log("creating test connection");
  connection = await createDbConnection("testConnection");
});

afterAll(async () => {
  console.log("closing test connection");
  await connection.close();
});

describe("Queries", () => {
  // Applies only to tests in this describe block

  it("fetches list of cards", async () => {
    // create an instance of ApolloServer that mocks out context, while reusing
    // existing dataSources, resolvers, and typeDefs.
    const { apolloServer, cardAPI } = await constructTestServer(connection, {
      context: () => ({})
    });

    // mock the datasources' underlying fetch methods
    cardAPI.getCards = jest.fn(async () =>
      Promise.resolve([mockFirstCardResponse])
    );

    // use our test server as input to the createTestClient fn
    // This will give us an interface, similar to apolloClient.query
    // to run queries against our instance of ApolloServer
    const { query } = createTestClient(apolloServer);
    const res = await query({
      query: GET_CARDS
    });
    expect(res).toMatchSnapshot();
  });

  it("fetches single card", async () => {
    const { apolloServer, cardAPI } = await constructTestServer(connection, {
      context: () => ({})
    });

    cardAPI.getCard = jest.fn(async () =>
      Promise.resolve(mockFirstCardResponse)
    );

    const { query } = createTestClient(apolloServer);
    const res = await query({
      query: GET_CARD,
      variables: { id: mockFirstCardResponseId }
    });
    expect(res).toMatchSnapshot();
  });
});

// describe("Mutations", () => {
//   it("create card", async () => {
//     const { server } = constructTestServer({
//       context: () => ({})
//     });

//     mockRepos.cards.save.mockReturnValueOnce(mockReturnCard);
//     const { mutate } = createTestClient(server);
//     const res = await mutate({
//       mutation: ADD_CARD,
//       variables: { input: mockCardInput }
//     });
//     expect(res).toMatchSnapshot();
//   });

//   it("update card", async () => {
//     const { server } = constructTestServer({
//       context: () => ({})
//     });

//     mockRepos.cards.findOne.mockReturnValueOnce(mockReturnCard);
//     mockRepos.cards.save.mockReturnValueOnce(mockReturnCard);
//     const { mutate } = createTestClient(server);
//     const res = await mutate({
//       mutation: UPDATE_CARD,
//       variables: { id: mockFirstCardResponseId, input: mockCardInput }
//     });
//     expect(res).toMatchSnapshot();
//   });

//   it("remove card", async () => {
//     const { server } = constructTestServer({
//       context: () => ({})
//     });

//     mockRepos.cards.remove.mockReturnValueOnce(mockReturnCard);
//     const { mutate } = createTestClient(server);
//     const res = await mutate({
//       mutation: REMOVE_CARD,
//       variables: { id: mockFirstCardResponseId }
//     });
//     expect(res).toMatchSnapshot();
//   });
// });
