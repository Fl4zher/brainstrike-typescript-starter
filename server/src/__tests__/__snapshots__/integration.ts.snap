// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutations create card 1`] = `
Object {
  "data": Object {
    "addCard": Object {
      "card": Object {
        "__typename": "Card",
        "created": "2020-02-01T20:19:27.909Z",
        "description": "testing",
        "id": "ZTMwM2Y3ZTUtMTVjZS00YTVmLTkxNzktYTc1ZmZiNGY4MTkxOkNhcmQ=",
        "label": "Test Card 1",
        "number": 1,
        "updated": null,
      },
      "message": "Card Added",
      "success": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutations remove card 1`] = `
Object {
  "data": Object {
    "removeCard": Object {
      "card": Object {
        "__typename": "Card",
        "created": "2020-02-01T20:19:27.909Z",
        "description": "testing",
        "id": "ZTMwM2Y3ZTUtMTVjZS00YTVmLTkxNzktYTc1ZmZiNGY4MTkxOkNhcmQ=",
        "label": "Test Card 1",
        "number": 1,
        "updated": null,
      },
      "message": "Card Removed",
      "success": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutations update card 1`] = `
Object {
  "data": Object {
    "updateCard": Object {
      "card": Object {
        "__typename": "Card",
        "created": "2020-02-01T20:19:27.909Z",
        "description": "testing",
        "id": "ZTMwM2Y3ZTUtMTVjZS00YTVmLTkxNzktYTc1ZmZiNGY4MTkxOkNhcmQ=",
        "label": "Test Card 1",
        "number": 1,
        "updated": null,
      },
      "message": "Card Updated",
      "success": true,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches list of cards 1`] = `
Object {
  "data": Object {
    "cards": Object {
      "edges": Array [
        Object {
          "cursor": "Q3xjYXJkfDc1Zjc3NDczLWI3NGMtNGEyMC1hZTBkLTczZTA0YjQyN2NkZQ==",
          "node": Object {
            "__typename": "Card",
            "created": "2020-02-01T20:19:27.909Z",
            "description": "Amet autem et.",
            "id": "NzVmNzc0NzMtYjc0Yy00YTIwLWFlMGQtNzNlMDRiNDI3Y2RlOkNhcmQ=",
            "label": "sed possimus tempore",
            "number": 99999,
            "updated": "2020-02-01T20:19:27.909Z",
          },
        },
        Object {
          "cursor": "Q3xjYXJkfDc4NTU1YTkwLWY2OGItNDU0Yy1iZWIyLWU3MDIxZjNkOGM2Mw==",
          "node": Object {
            "__typename": "Card",
            "created": "2020-02-01T23:03:03.942Z",
            "description": "Odit officia esse hic qui est perspiciatis.",
            "id": "Nzg1NTVhOTAtZjY4Yi00NTRjLWJlYjItZTcwMjFmM2Q4YzYzOkNhcmQ=",
            "label": "necessitatibus id error",
            "number": 99995,
            "updated": "2020-02-01T23:03:03.942Z",
          },
        },
        Object {
          "cursor": "Q3xjYXJkfGNhZjM4YjMxLWU1NTEtNDMzYi1iMWIwLWM2NTgwN2U5NzkyNw==",
          "node": Object {
            "__typename": "Card",
            "created": "2020-02-02T04:59:07.272Z",
            "description": "Et in sequi eum enim ratione.",
            "id": "Y2FmMzhiMzEtZTU1MS00MzNiLWIxYjAtYzY1ODA3ZTk3OTI3OkNhcmQ=",
            "label": "doloribus explicabo explicabo",
            "number": 99974,
            "updated": "2020-02-02T04:59:07.272Z",
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "Q3xjYXJkfGNhZjM4YjMxLWU1NTEtNDMzYi1iMWIwLWM2NTgwN2U5NzkyNw==",
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": "Q3xjYXJkfDc1Zjc3NDczLWI3NGMtNGEyMC1hZTBkLTczZTA0YjQyN2NkZQ==",
        "totalCount": 10000,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries fetches single card 1`] = `
Object {
  "data": Object {
    "card": Object {
      "__typename": "Card",
      "created": "2020-02-01T20:19:27.909Z",
      "description": "testing",
      "id": "ZTMwM2Y3ZTUtMTVjZS00YTVmLTkxNzktYTc1ZmZiNGY4MTkxOkNhcmQ=",
      "label": "Test Card 1",
      "number": 1,
      "updated": "2020-02-01T20:19:27.909Z",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
